### PR TITLE
Keycloakify v10 (Keycloak v24 & 25 support and much more)

### PR DESCRIPTION
**Pull request by garronej**
Complete overhaul of the React component library to match the latest bultin theme of Keycloak 24

### CODE REVIEW

**Comment by garronej** [comment]
I think I might be able to pull of retrocompatibility in the end.  
I do not guarenty it but it looks like it might be acheivable to still generate themes that works across all versions of Keycloak.

**Comment by garronej** [comment]
The new multi valued attribute requires some work to get validated in realtime on the frontend side:  

image

I'm on it

**Comment by garronej** [comment]
Thanks to https://github.com/jcputney/keycloak-theme-additional-info-extension we will finaly have proper clientside password validation let's go!  

For reference: https://keycloak.discourse.group/t/make-password-policies-available-to-freemarker/11632

**Comment by garronej** [comment]
I made good progress this week end but it will be at least another week befor this get merged

**Comment by garronej** [comment]
It's taking some time because I need to implement the logic for real-time client-side validation of user profile attributes. The default Keycloak theme lacks any form of client-side validation.

Trust me, the wait will be worth it for the enhanced user experience compared to the default Keycloak theme!

**Comment by garronej** [comment]
Keeping track of pages that have been updated:  

- [x] delete-credential.ftl
- [ ] login-update-profile.ftl
- [ ] login.ftl
- [ ] register.ftl
- [ ] select-authenticator.ftl
- [ ] templat.ftl
- [ ] update-email.ftl
- [ ] user-profile-commons.ftl
- [ ] webauthn-autenticate.ftl
- [ ] webauthn-register.ftl

**Comment by garronej** [comment]
**every** lpages of the login theme are now implemented by Keycloakify and up to date with Keycloak v24 🎊

**Comment by garronej** [comment]
We're getting there for real.  
image

**Comment by garronej** [comment]
Apologies for the delay, everyone. I'm working hard to get this shipped as soon as possible. There are still some crucial performance optimizations that need to be addressed. Currently, the generated JAR file is too large, and the build time is too long.

Despite these issues, this release is very solid, and I believe the time saved by the various improvements will make up for the delay. Thank you for your patience and understanding.

**Comment by garronej** [comment]
This should be my final update before the release, if all goes as planned. Given that this major update unavoidably includes some breaking changes, I've decided to implement all the significant modifications we've been considering for a while now. This has resulted in a bit of feature creep but we're getting there.

**Comment by garronej** [comment]
The new starter: https://github.com/keycloakify/keycloakify-starter/tree/keycloakify_10  

Full update of the documentation website and migrating guide coming up very soon.

**Comment by garronej** [comment]
Keycloak 25 just droped a few minut ago but don't worry it's confirmed working.

**Comment by garronej** [comment]
Keycloak 25 just droped.  
No issue but I just need to wait for [this PR](https://github.com/p2-inc/keycloak-account-v1/pull/13) to be merged.

**Comment by garronej** [comment]
Thank you @corneliusroemer!  
Well I'm indeed very exited about this release I think it will make devlopping and maintaining a Keycloak theme way easyer.  
But I'm also a bit embarassed because I told everyone that it would only take a week or two and it's been three month now, I ended up rewriting the all thing.

**Comment by rajshrimohanks** [comment]
> But I'm also a bit embarassed because I told everyone that it would only take a week or two and it's been three month now, I ended up rewriting the all thing (almost).

I'm pretty sure every developer out there knows and understands this. 😄 We are all grateful for your work @garronej. As someone who started using keycloakify when it was v6, I'm happy to see how it has evolved over time becoming more and more friendly to use. Thanks!

**Comment by garronej** [comment]
Thanks a lot for your kind word @rajshrimohanks !

**Comment by garronej** [comment]
We've put Keycloak v10 in production!  

https://filedn.com/lYPEPucNMheV19aDu23lK0Q/keycloakify/Demo-onyxia-theme.mp4

**Comment by garronej** [comment]
Tonight I updated the Webpack starter for the poor soul that cannot upgrade to Vite and for the avanturous pepoles that uses other web framworks:   

https://github.com/keycloakify/keycloakify-starter-cra

Also Keycloakify 10 has only two dependencies. [tsafe](https://github.com/garronej/tsafe) my TypeScript utilitiy toolkit and [react-markdown](https://github.com/remarkjs/react-markdown) for rendering the terms of services.  
This will allow you to use any package manager you want. I'll still keep `yarn` as the default since I think it's the more common but you'll be able to remove the `yarn.lock` and install with pnpm, bun, npm... without issues.

**Comment by garronej** [comment]
Today I worked on the Windows compat so you don't have to use WSL. My pain is immense. 

image

**Comment by garronej** [comment]
Status update: 

- Keycloakify 10 is ready for release
- The new starter for Vite and Webpack are ready for release.
- The new documentation website is 60% done.  
- The migration guide from v9 to v10 is not done yet.
- The Youtube tutorial isn't done yet.


If you can't wait anymore, you can start today:  
https://docs.keycloakify.dev/v/v10

I also published the new Storybook with all the components:  
https://storybook.keycloakify.dev/

**Comment by garronej** [comment]
I've solved the problems with fonts!  🥳
That's it, it just work. No gotchas, nothing specific to Keycloakify regarding importing font. You can import them as you would in any other web project!
https://docs.keycloakify.dev/custom-fonts

**Comment by garronej** [comment]
Writing the documentation is taking much longer than I anticipated.  
I'll post updates here on the pages I complete:  
- [Custom font](https://docs.keycloakify.dev/custom-fonts) It's just generic tutorial.
- [Importing assets](https://docs.keycloakify.dev/importing-assets) again, just a generic tutorial.  
- [Targetting Specific Keycloak Version](https://docs.keycloakify.dev/targetting-specific-keycloak-versions): To build only the jar for the Keycloak version you are using.  
- [i18n](https://docs.keycloakify.dev/i18n): It's nice, you can now define your user attributes messages in the Realm configuration.  See #549  
- [Terms and conditions](https://docs.keycloakify.dev/keycloak-configuration/terms-and-conditions) No more funny buisness with downloading Markdown files. It now works as with any other themes.  

To be continued...

**Comment by garronej** [comment]
Done with the Monorepo integration guide: https://docs.keycloakify.dev/keycloakify-in-my-app/as-a-subproject-of-your-monorepo

**Comment by garronej** [comment]
The Account REST API can be consumes from Keycloakify themes: https://docs.keycloakify.dev/account-rest-api

**Comment by jamesdh** [comment]
Hi @garronej! Potential first time keycloakify user here. We're in the process of upgrading our Keycloak from v22 to v25. The theming changes (and having to work in freemarker) are enough to make us finally look at keycloakify. With the release candidate do you think this in a solid enough state to give it a go?

**Comment by garronej** [comment]
Hello @jamesdh,

Nice to hear that!

Yes, we have Keycloakify 10 in production already, and many teams are using it. The main missing pieces are some pages of the documentation, and [the Account v3 REST API integration](https://docs.keycloakify.dev/account-rest-api) is still a bit rough around the edges.

I'll officially release Keycloakify when all the features are documented and I've recorded a general onboarding video tutorial. Unfortunately, writing documentation is way more time-consuming than I anticipated.

If you have questions, you can join our [Discord community](https://discord.gg/mJdYJSdcm4).

**Comment by jamesdh** [comment]
@garronej ok good to know! Wasn't sure as I gave the latest starter template a go within Keycloak docker (e.g. npx keycloakify start-keycloak) but it just spits out a blank page with the following error: 

> You've modified the configuration of the Keycloak server in a way that is incompatible with[this test application](https://github.com/keycloakify/my-theme.keycloakify.dev).

...and contains the following in the console: 
> [Error] Not allowed to request resource
[Error] Fetch API cannot load http://localhost:8080/realms/myrealm/.well-known/openid-configuration due to access control checks.
[Error] OIDC initialization error: Bad configuration. Timed out after 5000ms

Maybe something to look into? I couldn't see anything in the config that would cause that.

**Comment by garronej** [comment]
@jamesdh,  
Stranges 🤔,  

It's the first time it's reported to me.  
I just tried with a fresh starter:  

```bash
git clone https://github.com/keycloakify/keycloakify-starter
cd keycloakify-starter
yarn
npx keycloakify start-keycloak
# Select a version of Keycloak and reach https://my-theme.keycloakify.dev
```
It works with all the Keycloak version... 

Can you tell me what led to this message?

**Comment by jamesdh** [comment]
@garronej yea that's nearly the same exact thing I did, except I downloaded a zip of the source instead of checked it out since I was planning on moving it into an existing repo anyways. In any case, glad to know it's working for you! I'll dig around and try to find what might be causing this locally.

**Comment by jamesdh** [comment]
@garronej Ok, forgot to mention that I also ran `yarn build-keycloak-theme` before `npx keycloakify start-key cloak`. I just tried it on a different machine but am getting the same thing.

**Comment by garronej** [comment]
@jamesdh Maybe the port 8080 is already used by something else?  

Can you try with `npx keycloakify start-keycloak --port=8089` for example and tell me if it works.  

Otherwise, what OS are you using? what version of yarn (yarn --version)? I'll try to reproduce...

**Comment by jamesdh** [comment]
@garronej the docker container is correctly binding w/ 8080. Nothing else using it. Trying w/ 8089 results in the same. 

Since the package.json doesn't define a version for the package manager, I'm just using yarn 1.22.21. When package.json doesn't define a version that is the default that corepack installs. 

Do you want me to move this to another thread (issue), rather than pollute this PR?

**Comment by garronej** [comment]
@jamesdh yes please thank you :)  
The package manager shouldn't matter... are you on windows?

**Comment by garronej** [comment]
I've completed the documentations about integrating Keycloakify in an existing codebase.  

https://docs.keycloakify.dev/keycloakify-in-my-codebase  

We're getting there!

**Comment by corneliusroemer** [comment]
Most useful piece of documentation for most users (I suspect) will be the v9 to v10 migration guide.

While one can see how you changed the starter repo, there can be quite a few commits there and it's no super clear which are necessary and which aren't.

**Comment by garronej** [comment]
Theme Variants documentation completed: https://docs.keycloakify.dev/theme-variants

**Comment by garronej** [comment]
We're integrating Account v3 into Keycloakify!  

image

Currently, Keycloakify relies on Account v1. You can view the existing pages [here](https://storybook.keycloakify.dev/?path=/story/account-account-ftl--default). The new Account v3 is the one showed in the screenshot.

I've identified a way to provide a good development experience with the new system. If you're customizing the Account theme, it may be best to wait until the new system is released.

Thank you for your patience and support!

**Comment by venkata-ramana-subramanyam** [comment]
Hi @garronej , amazing work, I have one question and am not sure if I can ask it here or not.

In the latest RC, i see that there is a change in some of the types like LoginUpdateProfile

image

Earlier, it was like this

image


So, if I plan to make use of the latest RC version in my project, would I need to access the fields in the update profile like this? _"context.profile.attributesByName['email'].value"_

By the way, i see the keycloak context containing the _user_ object which has all the values. Could you pls provide your suggestion?

image

File: src/bin/shared/buildContext.ts [hunk 1]
```diff
@@ -0,0 +1,876 @@
...
+
+                assert(
+                    dirPath !== pathSep,
+                    "Couldn't find a place to run 'npm config get'"
+                );
+
+                try {
+                    child_process.execSync("npm config get", {
+                        cwd: dirPath,
+                        stdio: "ignore"
```
Comment (maxjoehnk) [inline src/bin/shared/buildContext.ts:578]: Setting this to `ignore` causes the stderr output not to be included in the thrown error which is why this code doesn't work as intended. Changing this to pipe fixes this issue without printing the output to the console.

```suggestion
                        stdio: "pipe"
```

Comment (garronej) [reply src/bin/shared/buildContext.ts]: Applied!

**Comment by garronej** [comment]
Hello @venkata-ramana-subramanyam,  

So yes, sorry this is a breaking change from Keycloak.  
In V9 you had `update-user-profile.ftl` that has been renamed `login-update-profile.ftl`.  

The legacy `login-update-profile.ftl` that you where implementing has been removed.  
But don't worry, you can reconstruct what was in the `kcContext.user` with what's in the kcContext.profile.attributesByNames.  

However maybe the best, to have all the validation is to simply start from scratch and reaply your changes to the new page.  

```bash
rm src/login/UpdateUserProfile.tsx
npx keycloakify eject-page # Select login -> update-user-profile.ftl
# Optional, if you want to customize the form at the component level
npm keycloakify eject-page # Select login -> UserProfileFormFields.ftl
```

**Comment by garronej** [comment]
I’ve finally completed the CSS level customization documentation:
https://docs.keycloakify.dev/customization-strategies/css-level-customization

In particular, some of you might want to check out the documentation specific to Tailwind: https://docs.keycloakify.dev/customization-strategies/css-level-customization/using-tailwind

**Comment by garronej** [comment]
https://github.com/user-attachments/assets/f8afa3ac-6884-4580-8184-00c941740f7f

Account theme v3 available soon!

**Comment by garronej** [comment]
Well I'm going to merge this branch even if I'm still not fully ready to release v10.  
It's kind of missleading v9 still being the main branch.  

For pepole waiting for the dust to have settled for upgrading you can track the official v10 release.

**Comment by garronej** [comment]
Hello!

Keycloakify v10 has been released.

Here are the migration guides: https://docs.keycloakify.dev/faq-and-help/migration-guides/v9-greater-than-v10

I've done my best, but I'm aware that this is not a smooth upgrade process. I figured that it was best to make all the breaking changes at once, hopefully avoiding the need for more in the future.

**Pull request merged**